### ISSUE TITLE
TYPE_USE annotations are not discovered by ASTHelpers

### ISSUE THREAD

**Issue by JeffreyFalgout** [google/error-prone#702]
Adding `PARAMETER` to the `@Target` declaration below, or changing `TYPE_USE` to `PARAMETER` will allow the below tests to pass. Currently, no annotation is returned if the target is only `TYPE_USE`. The annotation does appear in `VariableTree.getModifiers().getAnnotations()`, so it seems like `ASTHelpers` should be able to return it.

<hr>

````java
package name.falgout.jeffrey.experimental;

import com.google.errorprone.BugPattern;
import com.google.errorprone.BugPattern.SeverityLevel;
import com.google.errorprone.VisitorState;
import com.google.errorprone.bugpatterns.BugChecker;
import com.google.errorprone.bugpatterns.BugChecker.VariableTreeMatcher;
import com.google.errorprone.matchers.Description;
import com.google.errorprone.util.ASTHelpers;
import com.sun.source.tree.VariableTree;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@BugPattern(
    name = "TypeUseChecker",
    summary = "TYPE_USE annotations don't work.",
    severity = SeverityLevel.ERROR
)
public class TypeUseChecker extends BugChecker implements VariableTreeMatcher {
  @Retention(RetentionPolicy.SOURCE)
  @Target({ElementType.TYPE_USE})
  public @interface TypeUseAnnotation {
    boolean value();
  }

  @Override
  public Description matchVariable(VariableTree tree, VisitorState state) {
    TypeUseAnnotation annotation = ASTHelpers.getAnnotation(tree, TypeUseAnnotation.class);
    if (annotation != null && annotation.value()) {
      return buildDescription(tree).setMessage("TypeUseAnnotation was truthy.").build();
    }

    return Description.NO_MATCH;
  }
}
````

<hr>

````java
package name.falgout.jeffrey.experimental;

import com.google.errorprone.CompilationTestHelper;
import org.junit.Before;
import org.junit.Test;

public class TypeUseCheckerTest {
  private CompilationTestHelper compilationHelper;

  @Before
  public void setup() {
    compilationHelper = CompilationTestHelper.newInstance(TypeUseChecker.class, getClass());
  }

  @Test
  public void positiveCases(){
    compilationHelper.addSourceFile("PositiveCases.java").doTest();
  }

  @Test
  public void negativeCases() {
    compilationHelper.addSourceFile("NegativeCases.java").doTest();
  }
}
````

<hr>

````java
package name.falgout.jeffrey.experimental;

import name.falgout.jeffrey.experimental.TypeUseChecker.TypeUseAnnotation;

public class PositiveCases {
  // BUG: Diagnostic contains: truthy
  void consumeDataType(@TypeUseAnnotation(true) String string) {
  }
}
````

<hr>

````java
package name.falgout.jeffrey.experimental;

import name.falgout.jeffrey.experimental.TypeUseChecker.TypeUseAnnotation;

public class NegativeCases {
  void consumeDataType(@TypeUseAnnotation(false) String string) {
  }
}
````

**Comment by cushon** [comment]
`TYPE_USE` annotations annotate types, not declarations. `ASTHelpers.getAnnotation(tree, TypeUseAnnotation.class)` retrieves annotations from the variable declaration. To retrieve annotations from the variable's type instead, you can do something like `ASTHelpers.getType(tree).getAnnotationMirrors()`.

The annotation appears in `VariableTree.getModifiers().getAnnotations()` because there's no syntactic way for the parser to disambiguate whether `@A Foo f;` is a type annotation on `Foo` or a declaration annotation on `f` (or both); you have to wait until semantic information is available instead of looking at the AST.

**Comment by JeffreyFalgout** [comment]
`ASTHelpers.getType(tree).getAnnotationMirrors()` is an empty list. `ASTHelpers.getType(tree.getType()).getAnnotationMirrors()` is also an empty list. 

It looks like the way to get `TYPE_USE` annotations is `ASTHelpers.getType(tree).getRawTypeAttributes()`. Then you can pass the `AnnotationMirror` you get to `AnnotationProxyMaker::generateAnnotation` if you want an actual annotation instance.

<hr>

> ASTHelpers.getAnnotation(tree, TypeUseAnnotation.class) retrieves annotations from the variable declaration.

I think a more precise phrasing would be "retrieves annotations that apply to the variable declaration". As you say, the `TYPE_USE` annotation applies to the type of the variable declaration, but not the declaration itself.

Would you be willing to add `ASTHelpers::getTypeAnnotation` methods and update the documentation for `ASTHelpers::getAnnotation`? It was rather surprising to me that `ASTHelpers::getAnnotation` was returning `null` in this case (and the method's documentation didn't help).

I would be willing to create PR, if that helps.

**Comment by cushon** [comment]
> `ASTHelpers.getType(tree).getAnnotationMirrors()` is an empty list.
> It looks like the way to get `TYPE_USE` annotations is `ASTHelpers.getType(tree).getRawTypeAttributes()`

`getAnnotationMirrors` works for me. `getRawTypeAttributes` works too, but it's a method on `Symbol` not `Type`. Are you sure you were using `getType` instead of `getSymbol`?

**Comment by JeffreyFalgout** [comment]
Ah, I must've been using `getSymbol`. `getType` works for me as well. Interestingly, `getType().getAnnotation` always returns `null`, do you know why that is?

```
    @Override @DefinedBy(Api.LANGUAGE_MODEL)
    public <A extends Annotation> A getAnnotation(Class<A> annotationType) {
        return null;
    }
```

**Comment by cushon** [comment]
We now have a utility that abstracts over some of the quirks of the way type annotations are stored, which we're using the handle type annotations versions of `@Nullable`:

https://github.com/google/error-prone/blob/84b8cb398bf57f70b3af71e2b0f2c07cd9a8ac8b/check_api/src/main/java/com/google/errorprone/util/MoreAnnotations.java#L44
